### ISSUE TITLE
Make it more obvious when IDL are generated in the Model Explorer tab

### ISSUE THREAD

**Issue by J17359** [ZeligsoftDev/CX4CBDDS#286]
# **Issue and tracking information**
### Developer's Actual time spent on fix (hours)  
# **Issue reporter to provide a detailed description of the issue in the space below**
When you generate IDL from the Model Explorer tab, the idl files generated are only shown in the Project Explorer tab, it appears to the novice user that nothing has happened.

**Comment by eposse** [comment]
Hi @J17359. For this issue, would it suffice to put a "dirty" decorator on the model's root element icon? Or perhaps on the components for which IDL was generated?

**Comment by j26151** [comment]
I think 'dirty' decorator is a good idea regardless of this problem. I would like to see that if possible.

I think a solution of possibly outputting the the console "X IDL files generated, Y did not need to be updated" or something long those lines. Similar to how CX outputs to the console when validating and shows warnings/errors. This would allow us to see that something has at least happened. Same with generate descriptors.

**Comment by ysroh** [comment]
Can you log instead of "assert currentDisplay != null;"? Otherwise looks good.

**Comment by eposse** [comment]
Hi @j26151. Can you test the build from the "papyrus-issue286" branch? (pull request GH-291)

It will output to the console and the log, for all transformations (generation of IDL, descriptors and UML).

Let me know if this is sufficient to merge it.

Thanks

**Comment by ysroh** [comment]
@eposse I think this is a simple fix so you can merge it to the main branch.

**Comment by eposse** [comment]
@J17359 @j26151 I've merged this. It reports the results of generation (IDL, Descriptors and UML) to both the console and the log. Let me know if it is enough for you to close it.

**Comment by j26151** [comment]
@eposse What would you expect the output to look like? When I run this, I see that the console states that generation was successful for IDL on a given model, but it doesn't enumerate which files were created/updated and which were left alone

**Comment by eposse** [comment]
@j26151 I am able to generate a report of files that have been added, changed, removed and remained unchanged, but there is an problem: when using "Generate All" to run all generators (Descriptors, IDL, UML) we are unable to distinguish which generator produced or changed each file. We can report them in order (sorted by path for example) to make it easier to read, but to clearly identify which generator is responsible for each change has turned out to be a very difficult task, because of the way the generator engine and the individual generators were written, so implementing a report that properly discriminates the generated files by generator is quite difficult. Do you care about having such level of detail, or is it enough to report the combined lists of added, changed, removed and unchanged files (separate lists for each kind of update)?

**Comment by j26151** [comment]
@eposse Doesnt' Generate IDL, usually only generate IDL files (and taginfo.xml), Generate Descriptors generates .CDD and .CDP files, and Generate UML generates uml files?  We would be able to know who generated what based on that, right?

Even if the above isn't strictly true, we never use the "Generate All", because we (almost) never want to run Generate UML?  

Also, it would be nice if we could toggle/disable in the preferences the report for each of those options above "added, changed, removed, unchanged". I can see the number of unchanged being large for a number of our workflows and I could see a future ticket where our users do not want to see that information. It is good that we have access to it, but it would be great if we could toggle it.

**Comment by eposse** [comment]
And we should report to the Console and the Log, right? Or should we report somewhere else as well (a file, a pop up)?

**Comment by eposse** [comment]
I've fixed issues 1, 2 and 3 (although I haven't committed anything yet) and I have been able to reproduce item 4, although I am a bit confused about the folder structure of the generated folders. Nevertheless, I think it happens when you add a module as you say, but the new module doesn't need to have anything inside it. I'll need to dig into that code generator to understand why it is touching the other IDL, but my first thought is that it has to do with the confusing folder structure produced. For example, if the model contains a module SNA_Examples with a package "Package1", and inside that package there is a subpackage "defn" which contains, say a string "CXString1" typedef, then the generator will produce a folder called "Package1" with a subfolder "defn" and an idl file "SNA_Examples_defn.idl" inside "Package1" but not inside "defn" (?). If I add a module "CXModule1" to the "defn" package, the file "SNA_Examples_defn.idl" will be dirtied without any changes. The name of the generated file suggests to me that the generator will produce or update that file for any changes inside the UML Package element "defn", but since it is a module, it doesn't add anything to it. But the question is: don't you expect the module "CXModule1" to be nested in the "SNA_Examples" module in the IDL file?

**Comment by j26151** [comment]
In the attached image below, you will see that I added a module to the model which is then created in our SNA_Examples_defn.idl file. This behavior is what we would expect and we do not have any issues with how the model is generating IDL other than the fact that it constantly dirties that IDL each time.
![image](https://user-images.githubusercontent.com/49879646/138276455-0f8cf564-1ada-49e3-828e-0b21dfcae94e.png)

**Comment by j26151** [comment]
What do you mean? In my image, is the model that generated the IDL on the right. The entire process is working as we would expect.  NewModule exists within the defn package as well as within the SNA_Examples module. and the IDL conveys that.

**Comment by j26151** [comment]
If I take the model in my picture above, and I never change it again. I never modify the model itself again. I use as it exists in that picture. For all intents and purposes it is read only.

Then I run generate IDL on that model, the report will tell me that SNA_Examples_defn.idl has changed. And If I run generate IDL again, it again tell me that SNA_Examples_defn.idl has changed again. The model has not changed, but the report is telling me that the IDL has changed. That is the issue that I am talking about. I am not referring to any issue in the generated artifacts contents.

**Comment by eposse** [comment]
Ah! So successive calls to the generator on an unchanged model result in regenerating and dirtying the file even when nothing has changed. I think that was not clear to me from your original statement. 

And just to confirm: you see this behaviour only when having nested modules with typedefs in the same package? If it doesn't have nested modules you don't get the file dirty?

**Comment by eposse** [comment]
@J17359 @j26151 A status update: this is almost done, but while cleaning up the code I ran into an issue about how we collect and report files. I mentioned before that because of the way the generator is built, when you do a "Generate All", it's not trivial to separate the changes from each individual generator. I'm trying one last attempt at dealing with this, but if it doesn't work, we'll have to settle for a "combined" report. Finally, the last item was to filter the files reported so that we only see relevant files under the "unchanged" set. This also implied some additions to the individual code generators. 

I'm expecting to be done with this today.

**Comment by J17359** [comment]
> @J17359 @j26151 A status update: this is almost done, but while cleaning up the code I ran into an issue about how we collect and report files. I mentioned before that because of the way the generator is built, when you do a "Generate All", it's not trivial to separate the changes from each individual generator. I'm trying one last attempt at dealing with this, but if it doesn't work, we'll have to settle for a "combined" report. Finally, the last item was to filter the files reported so that we only see relevant files under the "unchanged" set. This also implied some additions to the individual code generators.
> 
> I'm expecting to be done with this today.

Thanks for the update @eposse . I'm sending an email to you all now summarizing what we think needs to be done prior to release.

**Comment by eposse** [comment]
@J17359 @j26151 I've just merged the fix. Would you like to test it before we tag it and build the release?

**Comment by j26151** [comment]
Every struct and DDS message now ends with the annotation "@top-level false" which breaks our build system and appears to be wrong for IDL provided to our AXCIOMA middleware. This is a bug that we would need fixed. Additionally, it looks like @elder4p added comments to the merged PR?

**Comment by eposse** [comment]
Hmm... That is strange. I don't know what this "@top-level false" tag comes from. I'm investigating.

I've addressed @elder4p comments with a new pull request #341.

**Comment by eposse** [comment]
In the generated idl I see
```
    @final
    struct TestData_msg {
```

but I do not see `@top-level false`.

Do you see this in BasicPubSub?

If not, which model?

**Comment by j26151** [comment]
![image](https://user-images.githubusercontent.com/49879646/140990502-563ed9e5-bcc1-4b68-9608-8e0c658b3cce.png)

At the end

**Comment by eposse** [comment]
Oh! I missed the comment. Thanks. I'm on it.

**Comment by eposse** [comment]
I've addressed the issue with the annotation and now it works again as it used to. However, in order to do that, we decided to revert some previous fix to the IDL generator which seems to be the cause of the wrong annotation. Unfortunately that fix we reverted was the one that fixed the issue of files being reported as "changed" when invoking the generator successively even though nothing changes in the model. Nevertheless, after testing a little bit, strangely enough, that problem seems to go away as well. So we are not 100% sure if the problem with unchanged files being reported as changed will pop-up again, but it is possible. This needs to be investigated further, but since it is not clear what the problem is exactly, and since we are making the release now, we decided this was the safest course of action, and we will investigate the problem in a new bug.
